File: scripts/test-interactive-wizard.mjs
import assert from "node:assert/strict";
import { execFileSync, spawnSync } from "node:child_process";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

const repoRoot = path.resolve(import.meta.dirname, "..");
const fixtureRoot = fs.mkdtempSync(path.join(os.tmpdir(), "create-stardrive-pty-"));
const fakeBin = path.join(fixtureRoot, "bin");
const projectName = "wizard-test";
const transcript = path.join(fixtureRoot, "transcript");

fs.mkdirSync(fakeBin);
fs.writeFileSync(
  path.join(fakeBin, "git"),
  `#!/bin/sh
case "$*" in
  *ls-remote*) printf 'deadbeef\trefs/tags/v1.5.9\\n' ;;
  *)
    for target; do :; done
    mkdir -p "$target/.git"
    printf '{"name":"stardrive","scripts":{}}\\n' > "$target/package.json"
    printf 'export default {\\n};\\n' > "$target/theme.config.ts"
    printf '#package-lock.json\\n' > "$target/.gitignore"
    ;;
esac
`,
  { mode: 0o755 },
);

execFileSync(process.execPath, ["build.mjs"], { cwd: repoRoot });

const command = `${JSON.stringify(process.execPath)} ${JSON.stringify(path.join(repoRoot, "bin/index.js"))} --version 1.5.9 --no-install`;
const args = process.platform === "darwin"
  ? ["-q", transcript, "sh", "-c", command]
  : ["-qefc", command, transcript];
const result = spawnSync("script", args, {
  cwd: fixtureRoot,
  env: {
    ...process.env,
    PATH: `${fakeBin}:${process.env.PATH}`,
    TERM: "xterm",
    npm_config_user_agent: "npm/11.17.0 node/v26.4.0 darwin x64",
  },
  input: `${projectName}\nn\nn\nn\nn\nn\n`,
  encoding: "utf8",
  timeout: 10_000,
});

assert.equal(result.error, undefined, `wizard timed out: ${result.error?.message ?? "unknown error"}`);
const output = fs.readFileSync(transcript, "utf8");
assert.equal(result.status, 0, `${result.stderr}\n${output}`);
for (const prompt of [
  "Project name:",
  "Keep the blog feature?",
  "Keep the FAQ feature?",
  "Keep the integration catalog?",
  "Keep the events feature?",
  "Will you host on Cloudflare Workers?",
]) {
  assert.match(output, new RegExp(prompt));
}
assert.match(output, /All systems go\./);
assert.match(
  fs.readFileSync(path.join(fixtureRoot, projectName, "theme.config.ts"), "utf8"),
  /droppedFeatures: \['blog', 'faq', 'integrations', 'events', 'cloudflare'\]/,
);

console.log("Interactive PTY wizard test passed");

File: scripts/test-interactive-wizard.py
#!/usr/bin/env python3
import os
import pty
import re
import select
import shutil
import subprocess
import sys
import tempfile
import time
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
PROJECT_NAME = "wizard-test"
PROMPTS = (
    ("Project name:", PROJECT_NAME),
    ("Keep the blog feature?", "n"),
    ("Keep the FAQ feature?", "n"),
    ("Keep the integration catalog?", "n"),
    ("Keep the events feature?", "n"),
    ("Will you host on Cloudflare Workers?", "n"),
)


def wait_for(master: int, output: bytearray, prompt: str, deadline: float) -> None:
    expected = prompt.encode()
    while expected not in output:
        remaining = deadline - time.monotonic()
        if remaining <= 0:
            raise AssertionError(f"timed out waiting for {prompt!r}:\n{output.decode(errors='replace')}")
        readable, _, _ = select.select([master], [], [], remaining)
        if not readable:
            continue
        try:
            chunk = os.read(master, 4096)
        except OSError:
            chunk = b""
        if chunk:
            output.extend(chunk)


def main() -> None:
    fixture_root = Path(tempfile.mkdtemp(prefix="create-stardrive-pty-"))
    try:
        fake_bin = fixture_root / "bin"
        fake_bin.mkdir()
        fake_git = fake_bin / "git"
        fake_git.write_text(
            "#!/bin/sh\n"
            "case \"$*\" in\n"
            "  *ls-remote*) printf 'deadbeef\\trefs/tags/v1.5.9\\n' ;;\n"
            "  *)\n"
            "    for target; do :; done\n"
            "    mkdir -p \"$target/.git\"\n"
            "    printf '{\"name\":\"stardrive\",\"scripts\":{}}\\n' > \"$target/package.json\"\n"
            "    printf 'export default {\\n};\\n' > \"$target/theme.config.ts\"\n"
            "    printf '#package-lock.json\\n' > \"$target/.gitignore\"\n"
            "    ;;\n"
            "esac\n"
        )
        fake_git.chmod(0o755)
        subprocess.run(["node", "build.mjs"], cwd=REPO_ROOT, check=True)

        master, slave = pty.openpty()
        env = os.environ | {
            "PATH": f"{fake_bin}:{os.environ['PATH']}",
            "TERM": "xterm",
            "npm_config_user_agent": "npm/11.17.0 node/v26.4.0 darwin x64",
        }
        process = subprocess.Popen(
            ["node", str(REPO_ROOT / "bin/index.js"), "--version", "1.5.9", "--no-install"],
            cwd=fixture_root,
            env=env,
            stdin=slave,
            stdout=slave,
            stderr=slave,
            close_fds=True,
        )
        os.close(slave)
        output = bytearray()
        deadline = time.monotonic() + 10
        for prompt, answer in PROMPTS:
            wait_for(master, output, prompt, deadline)
            os.write(master, f"{answer}\n".encode())
        while process.poll() is None:
            wait_for(master, output, "All systems go.", deadline)
        try:
            while True:
                chunk = os.read(master, 4096)
                if not chunk:
                    break
                output.extend(chunk)
        except OSError:
            pass
        finally:
            os.close(master)

        rendered = output.decode(errors="replace")
        assert process.returncode == 0, rendered
        for prompt, _ in PROMPTS:
            assert prompt in rendered, rendered
        assert "All systems go." in rendered, rendered
        theme = (fixture_root / PROJECT_NAME / "theme.config.ts").read_text()
        assert re.search(r"droppedFeatures: \['blog', 'faq', 'integrations', 'events', 'cloudflare'\]", theme), theme
        print("Interactive macOS PTY wizard test passed")
    finally:
        shutil.rmtree(fixture_root, ignore_errors=True)


if __name__ == "__main__":
    main()
